infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java

```diff
@@ -49,24 +45,54 @@ public boolean isServerReadyToHandleMsg(CorfuMsg msg) {
      * @param r   The router that took in the message.
      */
     public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        if (shutdown) {
+        if (getState() == ServerState.SHUTDOWN) {
             log.warn("Server received {} but is already shutdown.", msg.getMsgType().toString());
             return;
         }
 
+        if (!isServerReadyToHandleMsg(msg)) {
+            r.sendResponse(ctx, msg, CorfuMsgType.NOT_READY.msg());
+            return;
+        }
+
         if (!getHandler().handle(msg, ctx, r)) {
             log.warn("Received unhandled message type {}", msg.getMsgType());
         }
     }
 
+    protected void setState(ServerState newState) {
+        boolean updated = false;
+
+        while(!updated){
+            ServerState currState = getState();
+
+            if (currState == ServerState.SHUTDOWN && newState != ServerState.SHUTDOWN) {
+                throw new IllegalStateException("Server is in SHUTDOWN state. Can't be changed");
+            }
+
+            updated = state.compareAndSet(currState, newState);
```

Comment:
You can look up `updateAndGet`. It does the same thing just removes boilerplate code.

Reply:
updateAndGet doesn't work in my case, as you can see I make some extra checks.
Or if you have a good idea how it could be implemented, let me know

Reply:
Slava, you should be able to use updateAndGet:

```java
    public final V updateAndGet(UnaryOperator<V> updateFunction) {
        V prev, next;
        do {
            prev = get();
            next = updateFunction.apply(prev);
        } while (!compareAndSet(prev, next));
        return next;
    }
```

Reply:
fixed